### DISCUSSION TITLE
"Please update your system curl." but I have super-latest curl

### DISCUSSION THREAD

**Discussion by TonalidadeHidrica** [Homebrew/discussions#2107] [Getting started] [answered]
I'm trying to install linuxbrew to a Devuan machine which I don't have sudo permission.  Since the machine does not have curl installed, I manually compiled it from source and now placed at `~/.local/bin/curl`.  Then I tried to run the following script to install linuxbrew:

```
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

only to obtain the following error message.

```
Error: Please update your system curl.
Minimum required version: 7.41.0
Your curl version:
Your curl executable: /home/username/.local/bin/curl
Failed during: /home/username/.linuxbrew/bin/brew update --force --quiet
```

I definitely have the latest version of curl, as you can see below.  And actually linuxbrew *does* detect that curl is there, as shown in "Your curl executable".

```
$ which curl
/home/username/.local/bin/curl
$ curl --version
curl 7.78.0 (x86_64-pc-linux-gnu) libcurl/7.78.0 OpenSSL/3.0.0
Release-Date: 2021-07-21
Protocols: dict file ftp ftps gopher gophers http https imap imaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS HSTS HTTPS-proxy IPv6 Largefile NTLM NTLM_WB SSL TLS-SRP UnixSockets
```

But "Your curl version" is empty.  So I guess the only issue is that homebrew cannot somehow parse the `--version` output to obtain the correct version information.  How can I workaround?

**Comment by carlocab** [comment]
Try setting
```
export HOMEBREW_CURL=/path/to/curl
export HOMEBREW_DEVELOPER=1
```
Homebrew by default only looks for system-installed packages to use. There is experimental support for using a non-system `curl`, but it's gated by the `HOMEBREW_DEVELOPER` flag.

**Reply by TonalidadeHidrica** [reply]
Unfortunately it doesn't resolve the issue.

```
$ export HOMEBREW_CURL=/home/username/.local/bin/curl
$ export HOMEBREW_DEVELOPER=1
$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
==> Found cURL: /home/username/.local/bin/curl
==> Setting HOMEBREW_DEVELOPER to use Git/cURL not in /usr/bin
==> Checking for `sudo` access (which may request your password).
==> Select the Homebrew installation directory
- Enter your password to install to /home/linuxbrew/.linuxbrew (recommended)
- Press Control-D to install to /home/username/.linuxbrew
- Press Control-C to cancel installation
==> This script will install:
/home/username/.linuxbrew/bin/brew
/home/username/.linuxbrew/share/doc/homebrew
/home/username/.linuxbrew/share/man/man1/brew.1
/home/username/.linuxbrew/share/zsh/site-functions/_brew
/home/username/.linuxbrew/etc/bash_completion.d/brew
/home/username/.linuxbrew/Homebrew

Press RETURN to continue or any other key to abort
==> /bin/chown -R username:username /home/username/.linuxbrew/Homebrew
==> Downloading and installing Homebrew...
HEAD is now at 5c602003f Merge pull request #11930 from XuehaiPan/master
Error: Please update your system curl.
Minimum required version: 7.41.0
Your curl version:
Your curl executable: /home/username/.local/bin/curl
Failed during: /home/username/.linuxbrew/bin/brew update --force --quiet
$
```

And as I already stated, I think homebrew is aware that curl is at different location, since "Your curl executable" is pointing the right path.

**Reply by TonalidadeHidrica** [reply]
I have extracted the following small script [from here](https://github.com/Homebrew/brew/blob/364096b79ef58f91d29450a19b9dc6cc86c53624/Library/Homebrew/brew.sh#L423-L431) and ran, but it actually extracts the right curl version and does not output "oh, no!" so I have no idea what is wrong...

```bash
numeric() {
  # Condense the exploded argument into a single return value.
  # shellcheck disable=SC2086,SC2183
  printf "%01d%02d%02d%03d" ${1//[.rc]/ } 2>/dev/null
}

curl_version_output="$(${HOMEBREW_CURL} --version 2>/dev/null)"
curl_name_and_version="${curl_version_output%% (*}"

echo $curl_name_and_version

      message="Please update your system curl.
Minimum required version: ${HOMEBREW_MINIMUM_CURL_VERSION}
Your curl version: ${curl_name_and_version##* }
Your curl executable: $(type -p ${HOMEBREW_CURL})"
  echo $message

if [[ $(numeric "${curl_name_and_version##* }") -lt $(numeric "${HOMEBREW_MINIMUM_CURL_VERSION}") ]]
  then
        echo "oh, no!"
fi
```

```
curl 7.78.0
Please update your system curl. Minimum required version: Your curl version: 7.78.0 Your curl executable: /home/username/.local/bin/curl
```

**Reply by TonalidadeHidrica** [reply]
It was found out that `LD_LIBRARY_PATH` was no propagated to the script and `curl` failed with `/home/username/.local/bin/curl: error while loading shared libraries: libssl.so.3: cannot open shared object file: No such file or directory`.  So now I am wondering why it isn't.

**Reply by TonalidadeHidrica** [reply]
I found that the environment variables are filtered [here](https://github.com/Homebrew/brew/blob/364096b79ef58f91d29450a19b9dc6cc86c53624/bin/brew#L108-L118).  I don't know what is the purpose of the filtering...

**Reply by TonalidadeHidrica** [reply]
... so I succeeded in installing by exporting `HOMEBREW_NO_ENV_FILTERING=1`.  But it warns a lot that it is deprecated.  Why it is discouraged, and what is the better workaround?

**Reply by TonalidadeHidrica** [reply]
May be related: #932

**Reply by carlocab** [reply]
> ... so I succeeded in installing by exporting `HOMEBREW_NO_ENV_FILTERING=1`. But it warns a lot that it is deprecated. Why it is discouraged, and what is the better workaround?

The reason it is discouraged is in the [error message](https://github.com/Homebrew/brew/blob/364096b79ef58f91d29450a19b9dc6cc86c53624/bin/brew#L122):

```
Warning: HOMEBREW_NO_ENV_FILTERING is undocumented, deprecated and will be removed in a future Homebrew release (because it breaks many things)!
```

A better workaround would be to either 1) upgrade your system `curl` or 2) fix your updated non-system `curl` so that you don't need to set `LD_LIBRARY_PATH`.

You might also not need to set `HOMEBREW_NO_ENV_FILTERING` permanently by setting `HOMEBREW_FORCE_BREWED_CURL` instead. This will make Homebrew install `curl` for its own use. You'll still want to keep a relatively updated version of `curl` around as a backup though.

**Reply by carlocab** [reply]
@sjackman might also have input here.

**Reply by Bo98** [reply]
If `LD_LIBRARY_PATH` is necessary, the linkage of your `curl` installation seems a bit broken, which `ldd` will likely indicate.

**Reply by TonalidadeHidrica** [reply]
@carlocab Thanks a lot.  I'm currently trying to install `curl` and eventually force brew to use it.

@Bo98 Thanks for your advice.  Without `LD_LIBRARY_PATH`, `ldd ~/.local/bin/curl` shows this:

```
ldd ~/.local/bin/curl
        linux-vdso.so.1 (0x00007fff2c32b000)
        libcurl.so.4 => /home/username/.local/lib/libcurl.so.4 (0x00007f55e6fca000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f55e6fa2000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f55e6de1000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f55e70b6000)
        libssl.so.3 => not found
        libcrypto.so.3 => not found
```

while when `LD_LIBRARY_PATH` is present, all library is located properly:

```
        linux-vdso.so.1 (0x00007ffcd29d9000)
        libcurl.so.4 => /home/username/.local/lib/libcurl.so.4 (0x00007fc30f392000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fc30f36a000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc30f1a9000)
        libssl.so.3 => /home/username/.local/lib64/libssl.so.3 (0x00007fc30f101000)
        libcrypto.so.3 => /home/username/.local/lib64/libcrypto.so.3 (0x00007fc30eca1000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fc30f47e000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fc30ec9a000)
```

I guess this means that `curl` is dynamically linked to `libssl` and `libcrypto`.  Should I somehow link them statically?

**Reply by Bo98** [reply]
Curl was probably expecting them in `/home/username/.local/lib` but you have them installed in `/home/username/.local/lib64` - just a mismatch in install directories between how OpenSSL was installed and how Curl was installed.

**Reply by TonalidadeHidrica** [reply]
Ah, I see.  I installed both of them from source with the same `--prefix=` option, but it is now different.  I don't know why.  But maybe I should double-check the process.  Thanks a lot!

**Reply by carlocab** [reply]
Usually the `--libdir=` flag helps. You'll probably want to keep a working non-Homebrew `curl` around. I've on occasion broken my Homebrew `curl` installation (always my fault), but fixing it often requires a working `curl`.

**Comment by TonalidadeHidrica** [comment, marked as answer]
It worked!  So the solution is as follows.

* First, try installing `curl` with homebrew.  If you have `LD_LIBRARY_PATH` issue, you may temporarily set `export HOMEBREW_NO_ENV_FILTERING=1` variable.
* Once you succeeded in installing, set `export HOMEBREW_FORCE_BREWED_CURL=1`.  Then there will be no warning, no errors.  Maybe we may write it in `.profile`.
